### ISSUE TITLE
Potential cni e2e-simple breakage due to app http probe rewrite

### ISSUE THREAD

**Issue by incfly** [istio/istio#10557]
https://github.com/istio/istio/pull/10344

@tiswanso 

> This PR is causing e2e-simple to fail when run with auto-sidecar injection. Seems like the new helm chart param sidecarInjectorWebhook.rewriteAppHTTPProbe should default to true (maybe only when auth is enabled?).
> 
> The istio-cni release-1.1 nightly is failing due to it.

> To add info on what I've seen with this change... I think this has introduced a delta (at least for e2e-simple) for the e2e test framework run with sidecar injection enabled v. the default with istioctl kube-inject. I don't see how the e2e-simple would pass without the rewriteAppHTTPProbe=true so, I'm suspecting that the e2e test is running with it set for the istioctl kube-inject case. IMO, that delta in behavior is really bad and should be fixed by either changing the default for the flag or by having the makefile set it all the time for the e2e tests.

**Comment by tiswanso** [comment]
I think pilot/pkg/kube/inject/app_probe.go L51-53 is what changed the default behavior--at least compared to master. I'm not familiar with what was happening for the liveness probes prior to the existence of app_probe.go.

**Comment by incfly** [comment]
`RewriteAppHTTPProbe` should be default to false right? We only change probe when it's true.

How to reproduce the failure you mentioned?

**Comment by incfly** [comment]
@tiswanso mentioned the delta is about release-1.1 vs master. Not the delta before and after this feature.

Also it's e2e failed when the feature is turned off, which is really weird since we shouldn't depend on this feature to be on to have existing test pass. We need to understand that part.

**Comment by tiswanso** [comment]
The issue with e2e-simple is seen when run with sidecar-injector e.g:

```
E2E_ARGS="--kube_inject_configmap=istio-sidecar-injector" make test/local/auth/e2e_simple
```
I see the issue on baremetal and on CircleCI with kubeadm-dind-clusters, even without Istio CNI enabled.  The e2e-simple passes if run like this:

```
EXTRA_HELM_SETTINGS="--set sidecarInjectorWebhook.rewriteAppHTTPProbe=true" E2E_ARGS="--kube_inject_configmap=istio-sidecar-injector" make test/local/auth/e2e_simple
```

The failures seen is due to a service crashlooping due to the liveness probe failure.  

- nginx-default-http-backend service is created by the e2e-simple test but the deployment goes into crashloop due to liveness probe failures
```
istio-system   pod/nginx-default-http-backend-7ff6d875f5-zxpmh                  1/2       CrashLoopBackOff   6          9m        10.244.1.13   kube-node-2
```

```
Name:           nginx-default-http-backend-7ff6d875f5-zxpmh
Namespace:      istio-system
Node:           kube-node-2/10.192.0.4
Start Time:     Sun, 16 Dec 2018 09:58:53 +0000
Labels:         app=nginx-default-http-backend
                pod-template-hash=3992843191
Annotations:    sidecar.istio.io/status={"version":"17db20f8cc7828c840cf703a266a113165a26f3fc4ddda7fc9ddd1083d2ade30","initContainers":null,"containers":["istio-proxy"],"volumes":["istio-envoy","istio-certs"],"imageP...
Status:         Running
IP:             10.244.1.13
Controlled By:  ReplicaSet/nginx-default-http-backend-7ff6d875f5
Containers:
  backend:
    Container ID:   docker://4fa92c7ffda9377143ec0597f029aa82faf7ff755c959db82a49d9bcc1049de1
    Image:          gcr.io/google_containers/defaultbackend:1.4
    Image ID:       docker-pullable://gcr.io/google_containers/defaultbackend@sha256:865b0c35e6da393b8e80b7e3799f777572399a4cff047eb02a81fa6e7a48ed4b
    Port:           8080/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Sun, 16 Dec 2018 10:05:38 +0000
      Finished:     Sun, 16 Dec 2018 10:06:27 +0000
    Ready:          False
    Restart Count:  6
    Limits:
      cpu:     10m
      memory:  20Mi
    Requests:
      cpu:        10m
      memory:     20Mi
    Liveness:     http-get http://:8080/healthz delay=30s timeout=5s period=10s #success=1 #failure=3
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-787kc (ro)
  istio-proxy:
    Container ID:  docker://44243ea4521696dfaa0f6dc6948747703e730024f1dc97d78f5692e6d5096db4
    Image:         gcr.io/istio-release/proxyv2:release-1.1-latest-daily
    Image ID:      docker-pullable://gcr.io/istio-release/proxyv2@sha256:ae384e799ed478ce630c337a2522117137f689f3f885da4f13904b0295f809ca
    Port:          15090/TCP
    Host Port:     0/TCP
    Args:
      proxy
      sidecar
      --configPath
      /etc/istio/proxy
      --binaryPath
      /usr/local/bin/envoy
      --serviceCluster
      nginx-default-http-backend.default
      --drainDuration
      2s
      --parentShutdownDuration
      3s
      --discoveryAddress
      istio-pilot.istio-system:15011
      --zipkinAddress
      zipkin.istio-system:9411
      --connectTimeout
      1s
      --proxyAdminPort
      15000
      --controlPlaneAuthPolicy
      MUTUAL_TLS
      --statusPort
      15020
      --applicationPorts
      8080
    State:          Running
      Started:      Sun, 16 Dec 2018 09:59:08 +0000
    Ready:          True
    Restart Count:  0
    Requests:
      cpu:      10m
    Readiness:  http-get http://:15020/healthz/ready delay=1s timeout=1s period=2s #success=1 #failure=30
    Environment:
      POD_NAME:                      nginx-default-http-backend-7ff6d875f5-zxpmh (v1:metadata.name)
      POD_NAMESPACE:                 istio-system (v1:metadata.namespace)
      INSTANCE_IP:                    (v1:status.podIP)
      ISTIO_META_POD_NAME:           nginx-default-http-backend-7ff6d875f5-zxpmh (v1:metadata.name)
      ISTIO_META_CONFIG_NAMESPACE:   istio-system (v1:metadata.namespace)
      ISTIO_META_INTERCEPTION_MODE:  REDIRECT
      ISTIO_METAJSON_LABELS:         {"app":"nginx-default-http-backend"}

    Mounts:
      /etc/certs/ from istio-certs (ro)
      /etc/istio/proxy from istio-envoy (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-787kc (ro)
Conditions:
  Type           Status
  Initialized    True 
  Ready          False 
  PodScheduled   True 
Volumes:
  istio-envoy:
    Type:    EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:  Memory
  istio-certs:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  istio.default
    Optional:    true
  default-token-787kc:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-787kc
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason                 Age              From                  Message
  ----     ------                 ----             ----                  -------
  Normal   Scheduled              9m               default-scheduler     Successfully assigned nginx-default-http-backend-7ff6d875f5-zxpmh to kube-node-2
  Normal   SuccessfulMountVolume  9m               kubelet, kube-node-2  MountVolume.SetUp succeeded for volume "istio-envoy"
  Normal   SuccessfulMountVolume  9m               kubelet, kube-node-2  MountVolume.SetUp succeeded for volume "default-token-787kc"
  Normal   SuccessfulMountVolume  9m               kubelet, kube-node-2  MountVolume.SetUp succeeded for volume "istio-certs"
  Normal   Pulling                9m               kubelet, kube-node-2  pulling image "gcr.io/google_containers/defaultbackend:1.4"
  Normal   Pulled                 9m               kubelet, kube-node-2  Successfully pulled image "gcr.io/google_containers/defaultbackend:1.4"
  Normal   Pulled                 9m               kubelet, kube-node-2  Container image "gcr.io/istio-release/proxyv2:release-1.1-latest-daily" already present on machine
  Normal   Created                9m               kubelet, kube-node-2  Created container
  Normal   Started                9m               kubelet, kube-node-2  Started container
  Warning  Unhealthy              8m (x3 over 9m)  kubelet, kube-node-2  Readiness probe failed: HTTP probe failed with statuscode: 503
  Warning  Unhealthy              8m               kubelet, kube-node-2  Liveness probe failed: Get http://10.244.1.13:8080/healthz: read tcp 10.244.1.1:43860->10.244.1.13:8080: read: connection reset by peer
  Warning  Unhealthy              8m               kubelet, kube-node-2  Liveness probe failed: Get http://10.244.1.13:8080/healthz: read tcp 10.244.1.1:43966->10.244.1.13:8080: read: connection reset by peer
  Warning  Unhealthy              8m               kubelet, kube-node-2  Liveness probe failed: Get http://10.244.1.13:8080/healthz: read tcp 10.244.1.1:44074->10.244.1.13:8080: read: connection reset by peer
  Normal   Started                8m (x2 over 9m)  kubelet, kube-node-2  Started container
  Warning  Unhealthy              7m               kubelet, kube-node-2  Liveness probe failed: Get http://10.244.1.13:8080/healthz: read tcp 10.244.1.1:44502->10.244.1.13:8080: read: connection reset by peer
  Warning  Unhealthy              7m               kubelet, kube-node-2  Liveness probe failed: Get http://10.244.1.13:8080/healthz: read tcp 10.244.1.1:44606->10.244.1.13:8080: read: connection reset by peer
  Normal   Created                7m (x3 over 9m)  kubelet, kube-node-2  Created container
  Normal   Killing                7m (x2 over 8m)  kubelet, kube-node-2  Killing container with id docker://backend:Container failed liveness probe.. Container will be killed and recreated.
  Normal   Pulled                 7m (x2 over 8m)  kubelet, kube-node-2  Container image "gcr.io/google_containers/defaultbackend:1.4" already present on machine
  Warning  Unhealthy              4m (x4 over 5m)  kubelet, kube-node-2  (combined from similar events): Liveness probe failed: Get http://10.244.1.13:8080/healthz: read tcp 10.244.1.1:46628->10.244.1.13:8080: read: connection reset by peer
```

On a job run from istio.io/istio master the liveness probe HTTP port gets rewritten and the issue isn't seen.

**Comment by incfly** [comment]
Okay, now I understand. Let me try to summarize a bit.

1. I implemented the http probe rewrite to support mtls case, only for `istioctl kube-inject`.
1. Some auth tests start to depend on that. nginx container probing can work for auth enabled test.
1. I add a helm flag to not turn on this behavior by default.
1. The test depending on this feature starts to fail.

If so, given we don't want to turn on this feature by default for now, then we should figure out where we add test to depend on this feature, and revert that change. @tiswanso do you know where is the change on the test config side?

**Comment by incfly** [comment]
The option is to enable this feature on for auth related test. However, I prefer to simplify things and turn it off everywhere before we have e2e test for this feature.

**Comment by incfly** [comment]
I think I figured it out.

1. probe rewrite was checked in as #9266 
1. istio/cni repo run e2e_simple with auth enabled, https://github.com/istio/cni/pull/62

Should cni run with authn_enabled = false instead? Can we change that?

**Comment by tiswanso** [comment]
istio/cni could run with  rewriteAppHTTPProbe = true.  But my point is everyone running the e2e-simple with auth enabled needs that setting, regardless of CNI.  Is that the case as well for any mTLS enabled service with liveness probes?

I need to re-review the mTLS config in the system but is the proper fix one of the following:
  1. enable it in the `Makefile` generate_yaml istio-auth.yaml target?
  1. default the values.yaml of rewriteAppHTTPProbe to the value of global mTLS flag
  1. have the sidecar injector code check a flag for mTLS enabled on the service.

I'm thinking anyone else would hit this with auto-sidecar inject + mTLS services + liveness probes.

**Comment by incfly** [comment]
I prefer smallest needed change to get things unblocked. I'll send a PR to change https://github.com/istio/cni/blob/master/.circleci/run_e2e.sh with helm `--set sidecarInjectorWebhook.rewriteAppHTTPProbe=true`.

Once we have an e2e test we turn it on for istio-auth.yaml.

**Comment by incfly** [comment]
/cc @costinm 

Hold on. We found that e2e-simple tests did run with auth-enabled on prow, and they passed. For example, https://circleci.com/gh/istio/istio/272588?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

So we need to figure out why cni tests have to depend on this while the rest don't have to. Don't they also have the same nginx probe as well? There must be something different on the test runs.

**Comment by incfly** [comment]
@tiswanso I ran the test myself without specifying kube-inject's sidecar injector map, basically to simulate the e2e-simple with istio/istio run.

```
$ git checkout release-1.1 && E2E_ARGS=' -skip_cleanup=true' make test/local/auth/e2e_simple
NAME                                          READY     STATUS             RESTARTS   AGE
a-f4fb9d68f-cs7fs                             2/2       Running            0          8m
echosrv-deployment-1-59d9c8c8bb-7h9vt         2/2       Running            0          14m
echosrv-deployment-2-d6f5fd5c-whxxg           2/2       Running            0          14m
istio-citadel-7f4df674c-zpdmm                 1/1       Running            0          16m
istio-cleanup-secrets-1215a-rtc5d             0/1       Completed          0          16m
istio-egressgateway-654c9c65fb-gvmfp          1/1       Running            0          16m
istio-galley-858c7c58dd-76l8m                 1/1       Running            0          16m
istio-ingressgateway-785c996c47-nw69t         1/1       Running            0          16m
istio-pilot-769bcb6d9-lrghw                   2/2       Running            0          16m
istio-policy-569955585f-h4xpp                 2/2       Running            2          16m
istio-security-post-install-1215a-4wtq6       0/1       Completed          0          16m
istio-sidecar-injector-788d79d5f5-kbggc       1/1       Running            0          16m
istio-telemetry-655f7894fb-jqtnm              2/2       Running            2          16m
nginx-default-http-backend-56889fc4f6-hwz5h   1/2       CrashLoopBackOff   6          8m
nginx-ingress-controller-6d9c95f9f6-q9gnd     2/2       Running            0          8m
prometheus-7fcd5846db-nlwtc                   1/1       Running            0          16m
raw-cli-deployement-6f4459df-p8xpg            1/1       Running            0          14m
t-6769fd7bc6-wpm2c                            1/1       Running            0          8m

$ kc get deployment nginx-default-http-backend  -n istio-system -oyaml
https://gist.github.com/incfly/984196d9b155b6351ff249891d1ca2c5

```
You can see that the app container's probe is not modified, but the containers are still ready.

**Comment by incfly** [comment]
@tiswanso I ran the test myself without specifying kube-inject's sidecar injector map, basically to simulate the e2e-simple with istio/istio run.

What I found:
- The nginx application itself http probe is not rewritten, still at port 8080, original path. **rewriteAppProbe is not taking effect.**
- However, I did see the pod crash looping after a while and then came up.
- The test passed.

I think it's simply the race condition contribute to the flakiness that if `TestNginx` can run before kubelet think it's unhealthy, then it passed.

I think we should still flag guard feature but rewrite the testdata/nginx.yaml.tmp using curl or something else.

WDTY?
